strawberryfields/apps/train/parametrization.py

```diff
@@ -0,0 +1,363 @@
+    The initial value ``n_mean`` should be set high since varying :math:`\theta` can only lower
+    the mean number of clicks or photons.
+
+    The mapping from :math:`\theta` to :math:`W(\theta)` is specified using the ``embedding``
+    argument.
+
+    **Example usage:**
+
+    >>> g = nx.erdos_renyi_graph(4, 0.7, seed=1967)
+    >>> a = nx.to_numpy_array(g)
+    >>> embedding = Exp()  # TODO does this need updating?
+    >>> vgbs = VGBS(a, 3, embedding, True)
+    >>> params = np.array([0.05, 0.1, 0.02, 0.01])
+    >>> vgbs.A(params)
+    array([[0.        , 0.86070798, 0.93239382, 0.94176453],
+           [0.86070798, 0.        , 0.88692044, 0.89583414],
+           [0.93239382, 0.88692044, 0.        , 0.97044553],
+           [0.94176453, 0.89583414, 0.97044553, 0.        ]])
+    >>> vgbs.n_mean(params)
+    1.8906000271819803
+    >>> vgbs.generate_samples(params, 2)
+    array([[0, 0, 1, 1],
+           [1, 1, 0, 0]])
+
+    Args:
+        A (array): the input adjacency matrix :math:`A`
+        n_mean (float): the initial mean number of clicks or photons
+        embedding: the method of converting from trainable parameters :math:`\theta` to
+            :math:`W(\theta)` #TODO: any update from other PR?
+        threshold (bool): determines whether to use GBS in threshold or PNR mode
+        samples (array): an optional array of samples from :math:`A` used to speed up gradient
+            calculations #TODO: more info
+    """
+    def __init__(
+        self,
+        A: np.ndarray,
+        n_mean: float,
+        embedding,
+        threshold: bool,
+        samples: Optional[np.ndarray] = None,
+    ):
+        if not np.allclose(A, A.T):
+            raise ValueError("Input must be a NumPy array corresponding to a symmetric matrix")
+        self.A_init = A
+        self.embedding = embedding
+        self.threshold = threshold
+        self.A_init_scaled = rescale_adjacency(A, n_mean, threshold)
+        self.n_modes = len(A)
+        self._A_init_samples = None
+        if samples:
+            self._add_A_init_samples(samples)
+
+    def W(self, params: np.ndarray) -> np.ndarray:
+        """Calculate the diagonal matrix of weights :math:`W` that depends on the trainable
+        parameters :math:`\theta`.
+
+        **Example usage:**
+
+        >>> vgbs.W(params)
+        array([[0.95122942, 0.        , 0.        , 0.        ],
+               [0.        , 0.90483742, 0.        , 0.        ],
+               [0.        , 0.        , 0.98019867, 0.        ],
+               [0.        , 0.        , 0.        , 0.99004983]])
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            array: the diagonal matrix of weights
+        """
+        return np.diag(self.embedding(params))
+
+    @staticmethod
+    def _WAW(A: np.ndarray, W: np.ndarray) -> np.ndarray:
+        """Calculate the :math:`_WAW` parametrization.
```

Comment:
Maybe "calculate the WAW matrix?" I'm not sure a parametrization is something you calculate